File: packages/programming_using_c/grammar/words/js/token_stream.ts
import { IOstream } from '../../common/istream';
import { Token } from './token';

const END_OF_LINE = '';

const verbs = new Set(['rules', 'fly', 'swim']);
const nouns = new Set(['birds', 'fish', 'C++']);
const articles = new Set(['the', 'a']);
const prepositions = new Set(['and', 'or', 'but']);

export class Token_stream {
  private full = false;
  private buffer = Token.createEmpty();
  private istream: IOstream;

  constructor(istream: IOstream) {
    this.istream = istream;
  }

  putback(t: Token) {
    if (this.full) {
      throw new Error('try putback in full buffer');
    }
    this.full = true;
    this.buffer = t;
  }

  hasNest() {
    const t = this.get();
    if (t.isEmpty()) {
      return false;
    }
    this.putback(t);
    return true;
  }

  get() {
    if (this.full) {
      this.full = false;
      return this.buffer;
    }
    const char = this.istream.char();
    if (char === '.') {
      return Token.createEnd();
    }
    if (char === '') {
      return Token.createEmpty();
    }

    this.istream.putback(char);
    const word = this.istream.word();
    if (verbs.has(word)) {
      return Token.createVerb(word);
    }
    if (articles.has(word)) {
      return Token.createArticle();
    }
    if (nouns.has(word)) {
      return Token.createNoun(word);
    }
    if (prepositions.has(word)) {
      return Token.createPreposition(word);
    }
    throw new Error('Bad token');
  }
}

File: packages/programming_using_c/grammar/words/c++/token_stream.cpp
#include "include/std_lib_facilities.h"
#include <set>
#include "token.cpp"

std::set<std::string> verbs = {"rules", "fly", "swim"};
std::set<std::string> nouns = {"birds", "fish", "C++"};
std::set<std::string> articles = {"the", "a"};
std::set<std::string> prepositions = {"and", "or", "but"};

class Token_stream
{
public:
  Token get();
  void putback(Token t);

private:
  bool full{false};
  Token buffer = Token::createEmpty();
};

Token Token_stream::get() // read a token from cin
{
  if (full)
  {
    full = false;
    return buffer;
  }
  string word;
  cin >> word; // note that >> skips whitespace (space, newline, tab, etc.)

  if (word == "q")
  {
    return Token::createExit();
  }
  if (word == ".")
  {
    return Token::createEnd();
  }

  char last_char = word.at(word.length() - 1);
  if (last_char == '.')
  {
    cin.putback(last_char);
    word.pop_back();
  }
  if (verbs.contains(word))
  {
    return Token::createVerb(word);
  }
  if (nouns.contains(word))
  {
    return Token::createNoun(word);
  }
  if (articles.contains(word))
  {
    return Token::createArticle(word);
  }
  if (prepositions.contains(word))
  {
    return Token::createPreposition(word);
  }

  error("Bad token");
}

void Token_stream::putback(Token t)
{
  if (full)
  {
    error("putback() into a full buffer");
  }
  buffer = t;
  full = true;
}
